File: mobile/src/lib/taskAction.ts
import type { NexdoTask } from '../api/types';
import { parseServerDate } from './taskQuery';
import { detectTaskAction, type TaskActionIntent } from './taskActionDetector';
import { notificationDate, type TaskAction, type TaskActionStatus } from './todayActionQueue';

/**
 * `TaskAction`'s state machine, `TaskActionReconciler` and `TaskActionNotificationPlan`
 * (ios/Sources/NexdoCore/TaskAction.swift:21-115).
 *
 * The shape itself is already in `todayActionQueue.ts`, which Phase 4B ported; this adds the parts
 * that need the detector and the clock. All of it is pure, so all of it is tested without a device.
 */

export type { TaskAction, TaskActionStatus, TaskActionChannel } from './todayActionQueue';
export { notificationDate } from './todayActionQueue';

/** The extra fields the coordinator persists, beyond what the queue reads. */
export type StoredTaskAction = TaskAction & {
  type: TaskActionIntent;
  contactIdentifier?: string | null;
  context?: string | null;
  /** Epoch ms, set the moment an action enters `executing`. */
  executedAt?: number | null;
};

/**
 * `transition(to:now:)` (TaskAction.swift:41-63) —
 * "No communication can enter executing without an explicit approval event."
 *
 * Returns the new action and whether the move was allowed. Swift mutates in place and returns the
 * flag; this returns both so callers stay immutable.
 */
const ALLOWED: [TaskActionStatus, TaskActionStatus][] = [
  ['pending', 'scheduled'],
  ['scheduled', 'pending'],
  ['pending', 'awaitingApproval'],
  ['scheduled', 'awaitingApproval'],
  ['failed', 'awaitingApproval'],
  ['cancelled', 'awaitingApproval'],
  ['completed', 'awaitingApproval'],
  ['awaitingApproval', 'approved'],
  ['approved', 'executing'],
  ['executing', 'completed'],
  ['executing', 'failed'],
  ['approved', 'failed'],
  ['executing', 'awaitingApproval'],
  ['approved', 'awaitingApproval'],
  ['awaitingApproval', 'scheduled'],
  ['failed', 'scheduled'],
  ['cancelled', 'scheduled'],
  ['completed', 'scheduled'],
  ['executing', 'scheduled'],
];

export function transition(
  action: StoredTaskAction,
  next: TaskActionStatus,
  now: number,
): { action: StoredTaskAction; changed: boolean } {
  const listed = ALLOWED.some(([from, to]) => from === action.status && to === next);
  // "default: allowed = next == .cancelled || status == next" — anything may be cancelled, and a
  // no-op move to the current status always succeeds.
  const allowed = listed || next === 'cancelled' || action.status === next;
  if (!allowed) return { action, changed: false };

  return {
    action: { ...action, status: next, ...(next === 'executing' ? { executedAt: now } : {}) },
    changed: true,
  };
}

/** `requiresApproval` (TaskAction.swift:30) is a constant `true`; no channel is ever automatic. */
export const REQUIRES_APPROVAL = true;

/** A fresh action from a detection. `TaskAction.init` (TaskAction.swift:36-41). */
export function makeTaskAction({
  id,
  taskId,
  title,
  detection,
  scheduledAt,
}: {
  id: string;
  taskId: string;
  title: string;
  detection: NonNullable<ReturnType<typeof detectTaskAction>>;
  scheduledAt: number | null;
}): StoredTaskAction {
  return {
    id,
    taskId,
    sourceTitle: title,
    type: detection.intent,
    contactName: detection.contactName,
    preferredAction: detection.preferredAction,
    context: detection.context,
    scheduledAt,
    snoozedUntil: null,
    contactIdentifier: null,
    status: 'pending',
    executedAt: null,
  };
}

/**
 * `TaskActionReconciler.reconcile` (TaskAction.swift:67-88) —
 * "The saved task date wins over wording in an old title after rescheduling."
 *
 * One action per task at most, which is where the queue's one-action-per-task rule comes from: the
 * output is built by mapping over TASKS, so a second action for the same task cannot exist.
 */
export function reconcileActions({
  previous,
  tasks,
  now,
  timeZone,
  newId,
}: {
  previous: StoredTaskAction[];
  tasks: NexdoTask[];
  now: number;
  timeZone: string;
  /** Injected so tests are deterministic; the app passes `randomUUID`. */
  newId: () => string;
}): StoredTaskAction[] {
  const result: StoredTaskAction[] = [];

  for (const task of tasks) {
    if (task.status === 'COMPLETED' || task.status === 'CANCELLED') continue;
    const detected = detectTaskAction(task.title, now, timeZone);
    if (!detected) continue;

    const schedule = parseServerDate(task.startAt ?? task.dueAt ?? null);
    const old = previous.find((action) => action.taskId === task.id);

    // A different person or a different intent is a different action entirely.
    if (!old || old.contactName !== detected.contactName || old.type !== detected.intent) {
      result.push(makeTaskAction({ id: newId(), taskId: task.id, title: task.title, detection: detected, scheduledAt: schedule }));
      continue;
    }

    // Same person and intent, but the task moved or was retitled: rebuild, keeping the chosen contact.
    if ((old.scheduledAt ?? null) !== schedule || old.sourceTitle !== task.title) {
      const rebuilt = makeTaskAction({ id: newId(), taskId: task.id, title: task.title, detection: detected, scheduledAt: schedule });
      result.push({ ...rebuilt, contactIdentifier: old.contactIdentifier ?? null });
      continue;
    }

    result.push(old);
  }

  return result;
}

/** `TaskActionNotification` (TaskAction.swift:90-99). */
export type TaskActionNotification = {
  /** `"nexdo.action." + action.id` — the prefix the scheduler owns and replaces wholesale. */
  id: string;
  actionId: string;
  taskId: string;
  fireAt: number;
};

export const NOTIFICATION_ID_PREFIX = 'nexdo.action.';

/** The 48 Swift schedules at a time (TaskAction.swift:101-108). */
export const NOTIFICATION_LIMIT = 48;

/** `TaskActionNotificationPlan.desired` (TaskAction.swift:102-107). */
export function desiredNotifications(actions: StoredTaskAction[], now: number, limit = NOTIFICATION_LIMIT): TaskActionNotification[] {
  return actions
    .filter((action) => {
      if (action.status !== 'pending' && action.status !== 'scheduled') return false;
      const at = notificationDate(action);
      return at !== null && at > now;
    })
    .sort((left, right) => (notificationDate(left) as number) - (notificationDate(right) as number))
    .slice(0, Math.max(0, limit))
    .map((action) => ({
      id: NOTIFICATION_ID_PREFIX + action.id,
      actionId: action.id,
      taskId: action.taskId,
      fireAt: notificationDate(action) as number,
    }));
}

/** `TaskActionCoordinator.ownerKey(_:)` (TaskActionCoordinator.swift:43-45): SHA-256, lowercase hex. */
export async function ownerKey(userId: string, digest: (value: string) => Promise<string>): Promise<string> {
  return digest(userId);
}

File: ios/Sources/NexdoCore/TaskAction.swift
import Foundation

public enum TaskActionIntent: String, Codable, Sendable {
    case contact, call, message, email, followUp, schedule, openURL, navigate, book, remind
}
public enum TaskActionChannel: String, Codable, CaseIterable, Sendable { case call, message, email }
public enum TaskActionStatus: String, Codable, Sendable {
    case pending, scheduled, awaitingApproval, approved, executing, completed, failed, cancelled
}
public struct DetectedTaskAction: Equatable, Sendable {
    public var intent: TaskActionIntent
    public var contactName: String
    public var preferredAction: TaskActionChannel?
    public var scheduledAt: Date?
    public var context: String?
}
public protocol TaskActionDetector: Sendable {
    func detect(title: String, now: Date, timeZone: TimeZone) -> DetectedTaskAction?
}

public struct TaskAction: Codable, Equatable, Identifiable, Sendable {
    public var id: String
    public var taskId: String
    public var type: TaskActionIntent
    public var businessCandidateID: String?
    public var manualRecipient: TaskActionRecipient?
    public var contactIdentifier: String?
    public var contactName: String
    public var preferredAction: TaskActionChannel?
    public var scheduledAt: Date?
    public var context: String?
    public private(set) var status: TaskActionStatus = .pending
    public var requiresApproval: Bool { true }
    public private(set) var executedAt: Date?
    public var snoozedUntil: Date?
    public var sourceTitle: String
    public var notificationDate: Date? { snoozedUntil ?? scheduledAt }

    public init(taskId: String, title: String, detection: DetectedTaskAction, scheduledAt: Date?) {
        id = UUID().uuidString; self.taskId = taskId; sourceTitle = title
        type = detection.intent; contactName = detection.contactName
        preferredAction = detection.preferredAction; context = detection.context
        self.scheduledAt = scheduledAt
    }

    /// No communication can enter executing without an explicit approval event.
    @discardableResult public mutating func transition(to next: TaskActionStatus, now: Date = Date()) -> Bool {
        let allowed: Bool
        switch (status, next) {
        case (.pending, .scheduled), (.scheduled, .pending),
             (.pending, .awaitingApproval), (.scheduled, .awaitingApproval),
             (.failed, .awaitingApproval), (.cancelled, .awaitingApproval),
             (.completed, .awaitingApproval), (.awaitingApproval, .approved),
             (.approved, .executing), (.executing, .completed),
             (.executing, .failed), (.approved, .failed),
             (.executing, .awaitingApproval), (.approved, .awaitingApproval),
             (.awaitingApproval, .scheduled), (.failed, .scheduled),
             (.cancelled, .scheduled), (.completed, .scheduled), (.executing, .scheduled): allowed = true
        default: allowed = next == .cancelled || status == next
        }
        guard allowed else { return false }
        status = next
        if next == .executing { executedAt = now }
        return true
    }
}

public extension TaskAction {
    /// Opened before its reminder time (from a card or Today) and left without a choice. Nobody is choosing yet, so it
    /// goes back to `scheduled` and its reminder fires as planned. Once the time has arrived, `awaitingApproval` is the
    /// real state (the person is being asked now) and it stays.
    func isWaitingBeforeItsTime(now: Date) -> Bool {
        status == .awaitingApproval && (notificationDate.map { $0 > now } ?? false)
    }
    /// Returns an action waiting before its time to `scheduled`. True when it changed.
    @discardableResult mutating func releaseIfWaitingBeforeItsTime(now: Date) -> Bool {
        isWaitingBeforeItsTime(now: now) && transition(to: .scheduled, now: now)
    }
}

public enum TaskActionReconciler {
    /// Actions stuck waiting before their time go back to `scheduled`, except those whose Action screen is open.
    public static func release(_ actions: [TaskAction], except open: Set<String>, now: Date) -> [TaskAction] {
        actions.map { action in
            var value = action
            if !open.contains(action.id) { value.releaseIfWaitingBeforeItsTime(now: now) }
            return value
        }
    }
    /// The saved task date wins over wording in an old title after rescheduling.
    public static func reconcile(previous: [TaskAction], tasks: [NexdoTask], detector: any TaskActionDetector,
                                 now: Date, timeZone: TimeZone) -> [TaskAction] {
        tasks.compactMap { task in
            guard !task.isDone, task.status != "CANCELLED",
                  let detected = detector.detect(title: task.title, now: now, timeZone: timeZone) else { return nil }
            let schedule = (task.startAt ?? task.dueAt).flatMap(ServerDate.parse)
            guard var old = previous.first(where: { $0.taskId == task.id }),
                  old.contactName == detected.contactName, old.type == detected.intent else {
                return TaskAction(taskId: task.id, title: task.title, detection: detected, scheduledAt: schedule)
            }
            if old.scheduledAt != schedule || old.sourceTitle != task.title {
                let identifier = old.contactIdentifier
                let businessID = old.businessCandidateID
                let recipient = old.manualRecipient
                old = TaskAction(taskId: task.id, title: task.title, detection: detected, scheduledAt: schedule)
                old.contactIdentifier = identifier
                old.businessCandidateID = businessID
                old.manualRecipient = recipient
            }
            return old
        }
    }
}

public struct TaskActionNotification: Equatable, Sendable {
    public let id: String
    public let actionId: String
    public let taskId: String
    public let fireAt: Date
    public init?(action: TaskAction) {
        guard let date = action.notificationDate else { return nil }
        id = "nexdo.action." + action.id; actionId = action.id; taskId = action.taskId
        fireAt = date
    }
}
public enum TaskActionNotificationPlan {
    public static func desired(_ actions: [TaskAction], now: Date, limit: Int = 48) -> [TaskActionNotification] {
        actions.filter {
            ($0.status == .pending || $0.status == .scheduled) && ($0.notificationDate ?? .distantPast) > now
        }.sorted { $0.notificationDate! < $1.notificationDate! }.prefix(max(0, limit)).compactMap(TaskActionNotification.init)
    }
}
public protocol TaskActionNotificationScheduler: Sendable {
    func replace(with notifications: [TaskActionNotification], actions: [TaskAction], owner: String) async throws
}
